### ISSUE TITLE
Investigate what's needed for full Edge support

### ISSUE THREAD

**Issue by matatk** [matatk/landmarks#53]
Thanks to @menovak for trying this out already and showing it can be coaxed into working.

There have been a number of changes recently to this extension, and it also seems like Edge has come a long way with respect to WebExtension support, so really ought to try this and start logging and timetabling issues, for when it becomes possible to distribute extensions via the store.

**Comment by matatk** [comment]
Using [Microsoft's extension porting tool](https://docs.microsoft.com/en-us/microsoft-edge/extensions/guides/porting-chrome-extensions) I have got the extension running in Edge.  However, the pop-up immediately goes to an error state.  It is possible to inspect the popup page, but only by opening it in a separate tab—at which point I can't imagine there is any communication with the page "below" (as there isn't one).

I did try to inspect the popup page once I had it loaded in another tab, but couldn't find any errors.  This is a bit of a dead-end now; must try to contact them to ask about this, though...

Also: I don't think it's doing i18n on the keys in the manifest.  Must check if that is a known bug too.

**Comment by matatk** [comment]
Also, I got this error message from the porting tool (though don't think it's related to the above):

    options.js chrome-extension protocol should be replaced. recommend using getExtensionProtocol() - defined in API bridge

**Comment by menovak** [comment]
hi Matthew

For debugging an extension, I found this video the best source of how to:

https://docs.microsoft.com/en-us/microsoft-edge/extensions/guides/debugging-extensions

You can debug the menu - by setting breakpoints and then run it with the
entire page, so it does work...

can't speak to the i18n issue.

Mark

**Comment by menovak** [comment]
hi Matthew

I also had that same issue when using the example code provided by
Microsoft, and ended up replacing the text 'browser' with 'chrome' in I
think one or two places in my source code (don't ask me why..)

Marl

**Comment by matatk** [comment]
Thanks for the info @menovak.

A few things I've noticed:

 * It seems the [advice on the debugging page about there being an "inspect" option for the pop-up](https://docs.microsoft.com/en-us/microsoft-edge/extensions/guides/debugging-extensions#extension-page-debugging) isn't working: there's no "inspect" option, even after enabling developer stuff in `about:flags`. It's also not possible for me to right-click on an empty area of the popup. Will ask on the Microsoft Docs site.
 * The extension protocol warning was a false positive due to ['chrome-extension' being part of a URL in a comment in `options.js`](https://github.com/matatk/landmarks/blob/master/src/static/options.js#L32) :-).
 * Think I'll have to ask about the missing i18n in `package.json` too.

**Comment by matatk** [comment]
Updating to the Creators Update, as per the replies on the Microsoft docs site, should address this; nice.

**Comment by matatk** [comment]
Found [some info on SO about local file access](https://stackoverflow.com/a/39287930/1485308).

**Comment by matatk** [comment]
Was able to update to the latest version, and have discovered the following:

* Edge now uses the `browser` namespace (and not `msBrowser`—tested this empirically, though can't find any notes about the apparent change on the Edge documentation site.
* Therefore, the compatibility hook is really just for Chrome & Opera and should be removed from the others.
* Edge needs the `persistent` key in `background` in `manifest.json` to be set to `true`.
* Some extra [work will be needed for localisation](https://docs.microsoft.com/en-us/microsoft-edge/extensions/guides/packaging/localizing-extension-packages#localizing-name-and-description-in-the-windows-store), but will leave that for now. One thing that's a bit odd is that the extension's ID includes "\_\_MSG_extensionFullName__" so it looks like Edge is not behaving as expected there, but I've not been able to find out why.
* Edge doesn't support the `commands` API, so keyboard shortcuts can't be used at the moment. Bit of a showstopper, but could still make an Edge version so that it's there and ready for later on.
* Need to see if anyone can put extensions into the store yet.

**Comment by matatk** [comment]
After [trying some stuff](680b07485fc7d8d86e9892270093b1048537cb2c), the following occurred:

* Edge doesn't have NodeList.forEach, which other browsers have adopted. However, [they are working on it](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/10638731/) so a polyfill of some kind, that has an issue to remove it when possible, seems like the best approach (maybe a compatibility script for Edge?)
* The `command` stuff stops the execution of the background script, thus breaking everything else. Needs to be split into two smaller scripts, one of which can be omitted on Edge, for now? Is this worth it, as the keyboard shortcuts are rather important?
* I have assumed that the content script injector is necessary for Edge. Probably the best way to verify this is to continue along this line, but try to remove it later on, and see if things still work as expected.

**Comment by matatk** [comment]
#99 has added initial Edge support. Some observations:

* The extension does work, to a degree; at least on the first page that is open when it's loaded.
* Keyboard shortcuts are not supported yet [as above](https://github.com/matatk/landmarks/issues/53#issuecomment-329317016).
* Unfortunately there are several oddnesses that make it all seem a bit flaky at the moment:
  - Sometimes when reloading a page, one gets a `let/const redeclaration error` which seems to be a bug whereby some of the old pages's context is left behind.
  - Injection appears to work most of the time, e.g. mostly fine on http://matatk.agrip.org.uk/landmarks/ and http://bbc.co.uk/news/, but it doesn't work at all on <https://github.com/matatk/landmarks/>—no idea why.
* I did try to find out if injection was needed or not, but as it was flaky both ways, I can't be 100%, though it does look like it is needed.
* I realised that `compatibility.js` was the same for Chrome and Opera, and used a symlink but this breaks the build on Windows.

In summary: marking this as low-priority for now; will have to test the next major bump of Edge.

**Comment by matatk** [comment]
Tried with the new Chromium-based Edge and the Chrome Web Store version of the extension seems to work very well, and the process for submitting extensions is now much simpler—time to ship it officially :-).
